### PR TITLE
Issue1442 load coupling

### PR DESCRIPTION
**Pull request by AntoineGautier**
@mwetter Could you please review this PR to URBANopt staging branch? (The CI testing is limited to the package `Buildings.Applications` for now.)
It includes:

- base models: `Buildings.Applications.DHC.Loads.Examples.BaseClasses` and `Buildings.Applications.DHC.Networks.BaseClasses` 
- and validation and example models: `Buildings.Applications.DHC.Loads.Examples` and `Buildings.Applications.DHC.Loads.Validation`.

The full DHC system models based on your previous work on RN is in [issue1769_examples5GDHC](https://github.com/lbl-srg/modelica-buildings/tree/issue1769_examples5GDHC) but the development in that branch is not finalized yet, see #1769.

Eventually this branch has a dependency to `issue1129_energyPlus_zone` that will first be merged into `issue1437_district_heating_cooling` as discussed.

### CODE REVIEW

**Comment by AntoineGautier** [comment]
@mwetter I realized there are models from the wet coil development branch in that branch. I will close this PR and push a new clean branch. It that OK from your side?

**Comment by AntoineGautier** [comment]
@mwetter As discussed I:

- renamed the geoJSON export (and cleaned it as some files were not used in that branch),
- merged `issue1129_energyPlus_zone` into `issue1437_district_heating_cooling`.

You can start the review.

(I will open issues on `geojson-modelica-translator` for the rounding, air medium and building code name.)

File: Buildings/Applications/DHC/Loads/Validation/BaseClasses/FanCoil2PipeCooling.mo [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  Buildings.Controls.OBC.CDL.Continuous.Gain gaiHeaFlo1(k=1/QCoo_flow_nominal)
+    annotation (Placement(transformation(
+        extent={{-10,-10},{10,10}},
+        rotation=90,
+        origin={0,190})));
+equation
+  if have_fluPor then
+  end if;
+  if not have_QReq_flow then
+  end if;
```
Comment (mwetter) [inline Buildings/Applications/DHC/Loads/Validation/BaseClasses/FanCoil2PipeCooling.mo]: Empty if-then statements

File: Buildings/Applications/DHC/Loads/Validation/BaseClasses/FanCoil2PipeHeating.mo [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+<li>
+the air mass flow rate, from zero to its nominal value.
+</li>
+</ul>
+<p>
+The controller tracks the load while the impact of an unmet load on the room 
+air temperature is assessed with
+<a href=\"modelica://Buildings.Applications.DHC.Loads.BaseClasses.SimpleRoom\">
+Buildings.Applications.DHC.Loads.BaseClasses.SimpleRoom</a>.
+</p>
```
Comment (mwetter) [inline Buildings/Applications/DHC/Loads/Validation/BaseClasses/FanCoil2PipeHeating.mo]: Missing revision notes.

File: Buildings/Applications/DHC/Loads/Validation/BaseClasses/FanCoil2PipeHeatingValve.mo [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  Buildings.Controls.OBC.CDL.Logical.Sources.Constant con1(k=have_speVar)
+    annotation (Placement(transformation(extent={{-60,160},{-40,180}})));
+  Buildings.Controls.OBC.CDL.Logical.Switch swi
+    "Logical switch"
+    annotation (Placement(transformation(extent={{30,170},{50,190}})));
+equation
+  if have_fluPor then
+  end if;
+  if not have_QReq_flow then
+  end if;
```
Comment (mwetter) [inline Buildings/Applications/DHC/Loads/Validation/BaseClasses/FanCoil2PipeHeatingValve.mo]: Empty if-then

File: Buildings/Applications/DHC/Loads/Validation/BenchmarkFlowDistribution1.mo [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+</ul>
+</html>"),
+  Icon(coordinateSystem(preserveAspectRatio=false)), Diagram(
+        coordinateSystem(preserveAspectRatio=false, extent={{-120,-120},{120,
+            120}})),
+    experiment(
+      StopTime=2000000,
+      __Dymola_NumberOfIntervals=500,
+      Tolerance=1e-06,
+      __Dymola_Algorithm="Cvode"),
```
Comment (mwetter) [inline Buildings/Applications/DHC/Loads/Validation/BenchmarkFlowDistribution1.mo]: Remove __Dymola_NumberOfIntervals and __Dymola_Algorithm in all files.

File: Buildings/Applications/DHC/Loads/Validation/BenchmarkFlowDistribution1.mo [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    columns={2,3,4},
+    smoothness=Modelica.Blocks.Types.Smoothness.MonotoneContinuousDerivative1)
+    "Reader for thermal loads (y[1] is cooling load, y[2] is heating load)"
+    annotation (Placement(transformation(extent={{-100,-10},{-80,10}})));
+  Buildings.Controls.OBC.CDL.Continuous.Sources.Constant minTSet(k=20)
+    "Minimum temperature setpoint"
+    annotation (Placement(transformation(extent={{-100,60},{-80,80}})));
+  Buildings.Controls.OBC.UnitConversions.From_degC minTSet_K
+    "Minimum temperature setpoint (K)"
+    annotation (Placement(transformation(extent={{-60,60},{-40,80}})));
```
Comment (mwetter) [inline Buildings/Applications/DHC/Loads/Validation/BenchmarkFlowDistribution1.mo]: Please remove From_degC and use instead
```
minTSet(
  k(
  final unit="K",
  "displayUnit="degC") =293.15)
```

File: Buildings/Applications/DHC/Controls/MixingValveControl.mo [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+      annotation (Line(points={{11,20},{30,20},{30,0},{68,0}},
+               color={255,0,255}));
+    connect(zerCro.y, conTSup.trigger)
+      annotation (Line(points={{22,80},{40,80},{40,
+          40},{-80,40},{-80,-76},{-68,-76},{-68,-72}}, color={255,0,255}));
+  else
+    connect(fixMod.y, swi.u2)
+      annotation (Line(points={{11,0},{68,0}},
+                     color={255,0,255}));
+  end if;
```
[the remark this answers is not in the document]
Comment (AntoineGautier) [reply Buildings/Applications/DHC/Controls/MixingValveControl.mo]: You are right: not needed here.

File: Buildings/Applications/DHC/Loads/BaseClasses/FlowDistribution.mo [hunk 7]
```diff
@@ -0,0 +1,866 @@
...
+      connect(port_a, pumFlo.port_a)
+        annotation (Line(points={{-100,0},{-50,0}}, color={0,127,255}));
+      connect(port_a, pumSpe.port_a)
+        annotation (Line(points={{-100,0},{-56,0},{-56,-40},{-50,-40}},color={0,127,255}));
+    else
+      connect(port_a, pipPre.port_a)
+        annotation (Line(points={{-100,0},{-16,0}}, color={0,127,255}));
+    end if;
+
+  end if;
```
[the remark this answers is not in the document]
Comment (AntoineGautier) [reply Buildings/Applications/DHC/Loads/BaseClasses/FlowDistribution.mo]: You are right, I could clean up several equation sections. A few conditional statements are still needed for this reason: statements involving conditionally removed components are removed at translation time by Modelica specification. However, obsolete statements corresponding to the default model structure need to be programmatically removed. I included that comment in the equation section when needed.

**Comment by AntoineGautier** [comment]
@mwetter Thanks a lot for the detailed review. I addressed all your comments (here above and on paper). There are two remaining points to agree on.

- I wanted to add a validation case for the `PartialBuilding` configuration where `PHea` and `PCoo` are used, i.e., for decentralized heating and cooling systems like split systems. However, implementing a split model is not as straightforward as I expected (cf. our discussion by email). I still would like to merge the model as is because I do not see where the model could fail considering the simplicity of that option.
```
  parameter Boolean have_eleHea = true
    "Set to true if the building has decentralized electric heating equipment"
    annotation(Evaluate=true);
...
  Modelica.Blocks.Interfaces.RealOutput PHea(
    final quantity="Power", final unit="W") if have_eleHea
    "Power drawn by decentralized heating equipment"
```

- I agree that the logic for modeling a change-over terminal unit is not intuitive (`have_watCoo = true`, `have_chaOve = true`, and `have_watHea = false`). That is why I tried to document it thoroughly. The reason for that situation is the use of two pairs of fluid ports for heating and chilled water, instead of a fluid port array. One of these two pairs needs to be removed for a change-over system. Keeping the chilled water label makes more sense to me as, fundamentally, a change-over coil is a chilled water coil (with condensate draining) configured to also operate with heating water.

If we agree on that then this is ready to merge when all test pass.

**Comment by mwetter** [comment]
@AntoineGautier The first point (merging without a test for a split system) is fine, although we could approximate a sensible only split system with one of the `HeatPumps.Carnot_*`.

Let's discuss the second point on today's call. What I was thinking is an enumeration such as the one below (types are just examples)
```
  type HeatingCooling = enumeration(
      ChangeOverHydronicHydronicHeating "Change over system with DX coil and hydronic heating",
      ChangeOverReversibleHeatPump "Changer over system with reversible heat pump",
      HydronicHeatingOnly "Only hydronic heating",
      HydronicCoolingOnly "Only hydronic cooling")
```
Then, we would assign the `have_*` parameters based on this enumeration using `final`. Other systems could be added in the future as needed.

**Comment by AntoineGautier** [comment]
@mwetter This is ready to merge when all tests pass.

This closes #1442 and URBANopt [#35](https://github.com/urbanopt/urbanopt-modelica/issues/35).

**Comment by mwetter** [comment]
@AntoineGautier : I don't see why this one does not run the CI tests at https://github.com/lbl-srg/modelica-buildings/pull/1730 as another one from this morning runs: https://travis-ci.com/github/lbl-srg/modelica-buildings/builds/155572411

If you tested it locally, please feel free to merge.

**Pull request merged**